registration/registration.go

```diff
@@ -317,9 +371,46 @@ func (r *Registration) newRoundOpts() []newRoundOptionFunc {
 	}
 }
 
+func (r *Registration) verifyCert(
+	certificate *shared.OpaqueCert,
+	certPubKeyHint, nodeID []byte,
+) error {
+	var matchingKeys [][]byte
+
+	if len(certPubKeyHint) == 0 {
+		matchingKeys = append(matchingKeys, r.cfg.Certifier.PubKey.Bytes())
+	} else {
+		r.trustedKeysMtx.RLock()
+
+		certKeys := append(r.trustedPublicKeys, r.cfg.Certifier.PubKey.Bytes())
+		for _, certKey := range certKeys {
+			if len(certKey) >= len(certPubKeyHint) && bytes.Equal(certKey[:len(certPubKeyHint)], certPubKeyHint) {
+				matchingKeys = append(matchingKeys, certKey)
+			}
+		}
+
+		r.trustedKeysMtx.RUnlock()
+
+		if len(matchingKeys) == 0 {
+			return ErrNoMatchingCertPublicKeys
+		}
+	}
+
+	var certErr error
+	for _, key := range matchingKeys {
+		_, err := shared.VerifyCertificate(certificate, key, nodeID)
+		if err == nil {
+			// cert verified successfully
+			break
+		}
+		certErr = err
+	}
+	return certErr
```

Comment:
I'm not sure this code is correct. If multiple `matchingKeys` exist and the first fails to verify, but the second does then this method will return an error? How about this instead:

```suggestion
	for _, key := range matchingKeys {
		_, err := shared.VerifyCertificate(certificate, key, nodeID)
		if err == nil {
			return nil
		}
	}
	return errors.New("failed to verify certificate")
```